File: bdp-devops/ambari/ambari-common/src/main/python/resource_management/libraries/functions/mpack_version.py
"""
  Licensed to the Apache Software Foundation (ASF) under one
  or more contributor license agreements.  See the NOTICE file
  distributed with this work for additional information
  regarding copyright ownership.  The ASF licenses this file
  to you under the Apache License, Version 2.0 (the
  "License"); you may not use this file except in compliance
  with the License.  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
"""

import re

"""
 This class should be used to compare mpack and stack versions.
 Base method which should be used is parse/parse_stack_version, depends
 on which versions you want to compare. This method will validate and parse
 version which you will pass as parameter, and return object of current class with
 parsed version. Same thing you should do with another version, with which you are
 planning to compare previous one. After that, use "==", ">", "<" to get final result.
"""
class MpackVersion(object):
  __mpack_version_pattern = "(?P<major>[0-9]+).(?P<minor>[0-9]+).(?P<maint>[0-9]+)(-h(?P<hotfix>[0-9]+))*-b(?P<build>[0-9]+)"
  __mpack_legacy_stack_version_pattern = "(?P<major>[0-9]+).(?P<minor>[0-9]+).(?P<maint>[0-9]+).(?P<hotfix>[0-9]+)(-(?P<build>[0-9]+))"
  __mpack_version_regex = re.compile(__mpack_version_pattern)
  __mpack_legacy_stack_version_regex = re.compile(__mpack_legacy_stack_version_pattern)

  def __init__(self, major, minor, maint, hotfix, build):
    """
    :type major int
    :type minor int
    :type maint int
    :type hotfix int
    :type build int
    """
    self.__major = int(major)
    self.__minor = int(minor)
    self.__maint = int(maint)
    self.__hotfix = int(hotfix) if hotfix else 0  # hotfix is optional group
    self.__build = int(build)

  def __repr__(self):
    return "{0}.{1}.{2}-h{3}-b{4}".format(*self.to_list())

  def to_list(self):
    """
    Return version elements as list

    :rtype list
    """
    return [
      self.__major,
      self.__minor,
      self.__maint,
      self.__hotfix,
      self.__build
    ]

  def __cmp__(self, other):
    """
    :type other MpackVersion

    :raise TypeError
    """
    if other and not isinstance(other, self.__class__):
      raise TypeError("Operand type is different from {0}".format(self.__class__.__name__))

    r = 0
    x = self.to_list()
    y = other.to_list()

    for i in range(0, len(x)):
      r = x[i] - y[i]
      if r != 0:
        break

    return 1 if r > 0 else -1 if r < 0 else 0

  @classmethod
  def parse(cls, mpack_version):
    """
      Parse string to mpack version

      :type mpack_version str
      :rtype MpackVersion
      """
    matcher = cls.validate(mpack_version)
    return MpackVersion(
      matcher.group("major"),
      matcher.group("minor"),
      matcher.group("maint"),
      matcher.group("hotfix"),
      matcher.group("build")
    )


  @classmethod
  def parse_stack_version(cls, stack_version):
    """
      Parse string to mpack version

      :type stack_version str
      :rtype MpackVersion
      """
    matcher = cls.validate_stack_version(stack_version)
    return MpackVersion(
      matcher.group("major"),
      matcher.group("minor"),
      matcher.group("maint"),
      matcher.group("hotfix"),
      matcher.group("build")
    )


  @classmethod
  def validate_stack_version(cls, stack_version):
    """
    Check if provided version is valid. If version is valid will return match object
    or will raise exception.

    :param stack_version version to check
    :type stack_version str

    :rtype __Match[T] | None

    :raise ValueError
    """

    if not stack_version:
      raise ValueError("Module version can't be empty or null")

    version = stack_version.strip()

    if not version:
      raise ValueError("Module version can't be empty or null")

    matcher = cls.__mpack_version_regex.match(version)

    if not matcher:
      matcher = cls.__mpack_legacy_stack_version_regex.match(version)
      if not matcher:
        raise ValueError("{0} is not a valid {1}".format(version, cls.__name__))
    else:
      if not matcher.group("hotfix"):
        raise ValueError("{0} is not a valid {1}".format(version, cls.__name__))

    return matcher


  @classmethod
  def validate(cls, mpack_version):
    """
    Check if provided version is valid. If version is valid will return match object
    or will raise exception.

    :param module_version version to check
    :type module_version str

    :rtype __Match[T] | None

    :raise ValueError
    """

    if not mpack_version:
      raise ValueError("Module version can't be empty or null")

    version = mpack_version.strip()

    if not version:
      raise ValueError("Module version can't be empty or null")

    matcher = cls.__mpack_version_regex.match(version)

    if not matcher:
      raise ValueError("{0} is not a valid {1}".format(version, cls.__name__))

    return matcher

  @property
  def major(self):
    return self.__major

  @property
  def minor(self):
    return self.__minor

  @property
  def maint(self):
    return self.__maint

  @property
  def hotfix(self):
    return self.__hotfix

  @property
  def build(self):
    return self.__build

File: dataSRE/ambari-versions/ambari-3.0.0.0-SNAPSHOT/ambari-server/src/main/java/org/apache/ambari/server/utils/MpackVersion.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.ambari.server.utils;


import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.apache.commons.lang.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * This class should be used to compare mpack and stack versions.
 * Base method which should be used is parse/parseStackVersion, depends
 * on which versions you want to compare. This method will validate and parse
 * version which you will pass as parameter, and return object of current class with
 * parsed version. Same thing you should do with another version, with which you are
 * planning to compare previous one. After that, use method compare to get final result.
 */

public class MpackVersion implements Comparable<MpackVersion> {

  // RE for different version formats like N.N.N-hN-bN, N.N.N-bN, N.N.N.N-N
  private final static String VERSION_WITH_HOTFIX_AND_BUILD_PATTERN = "^([0-9]+).([0-9]+).([0-9]+)-h([0-9]+)-b([0-9]+)";
  private final static String VERSION_WITH_BUILD_PATTERN = "^([0-9]+).([0-9]+).([0-9]+)-b([0-9]+)";
  private final static String LEGACY_STACK_VERSION_PATTERN = "^([0-9]+).([0-9]+).([0-9]+).([0-9]+)-([0-9]+)";
  private final static String FORMAT_VERSION_PATTERN = "^([0-9]+)(\\.)?([0-9]+)?(\\.)?([0-9]+)?(-h)?([0-9]+)?(-b)?([0-9]+)?";

  // Patterns for previous RE
  private final static Pattern PATTERN_WITH_HOTFIX = Pattern.compile(VERSION_WITH_HOTFIX_AND_BUILD_PATTERN);
  private final static Pattern PATTERN_LEGACY_STACK_VERSION = Pattern.compile(LEGACY_STACK_VERSION_PATTERN);
  private final static Pattern PATTERN_WITHOUT_HOTFIX = Pattern.compile(VERSION_WITH_BUILD_PATTERN);
  private final static Pattern PATTERN_FORMAT_VERSION = Pattern.compile(FORMAT_VERSION_PATTERN);

  private final static Logger LOG = LoggerFactory.getLogger(MpackVersion.class);

  // Parts of version
  private int major;
  private int minor;
  private int maint;
  private int hotfix;
  private int build;


  public MpackVersion(int major, int minor, int maint, int hotfix, int build) {
    this.major = major;
    this.minor = minor;
    this.maint = maint;
    this.hotfix = hotfix;
    this.build = build;
  }

  /**
   * Method which will parse mpack version
   * which user passed as parameter. Also
   * in this method version will be validated.
   * @param mpackVersion string
   * @return MpackVersion instance which contains parsed version
   * */
  public static MpackVersion parse(String mpackVersion) {
    return parse(mpackVersion, true);
  }

  public static MpackVersion parse(String mpackVersion, boolean strict) {

    if(!strict) {
      mpackVersion = format(mpackVersion);
    }
    Matcher versionMatcher = validateMpackVersion(mpackVersion);
    if (versionMatcher.pattern().pattern().equals(VERSION_WITH_BUILD_PATTERN)) {
      return new MpackVersion(Integer.parseInt(versionMatcher.group(1)), Integer.parseInt(versionMatcher.group(2)),
              Integer.parseInt(versionMatcher.group(3)), 0, Integer.parseInt(versionMatcher.group(4)));
    } else if (versionMatcher.pattern().pattern().equals(VERSION_WITH_HOTFIX_AND_BUILD_PATTERN)) {
      return new MpackVersion(Integer.parseInt(versionMatcher.group(1)), Integer.parseInt(versionMatcher.group(2)),
              Integer.parseInt(versionMatcher.group(3)), Integer.parseInt(versionMatcher.group(4)), Integer.parseInt(versionMatcher.group(5)));
    } else {
      throw new IllegalArgumentException("Wrong format for mpack version");
    }
  }

  /**
   * Method to format an mpack version in {major}.{minor}.{maint}-h{hotfix}-b{build} format
   * @param mpackVersion input mpack version string
   * @return formatted mpack version string
   */
  public static String format(String mpackVersion) {
    Matcher m = PATTERN_FORMAT_VERSION.matcher(mpackVersion);
    if(m.matches()) {
      String majorVersion = m.group(1);
      String minorVersion = m.group(3);
      String maintVersion = m.group(5);
      String hotfixNum = m.group(7);
      String buildNum = m.group(9);
      if(hotfixNum != null || buildNum != null) {
        if(minorVersion == null || maintVersion == null) {
          // Both minorVersion and maintVersion should be specified
          throw new IllegalArgumentException("Wrong format for mpack version");
        }
      }
      minorVersion = minorVersion != null? minorVersion: "0";
      maintVersion = maintVersion != null? maintVersion: "0";
      hotfixNum = hotfixNum != null? hotfixNum: "0";
      buildNum = buildNum != null? buildNum: "0";
      String formattedMpackVersion = String.format("%s.%s.%s-h%s-b%s",
        majorVersion, minorVersion, maintVersion, hotfixNum, buildNum);
      return formattedMpackVersion;

    } else {
      throw new IllegalArgumentException("Wrong format for mpack version");
    }
  }

  /**
   * Method which will parse stack version
   * which user passed as parameter. Also
   * in this method version will be validated.
   * @param stackVersion string
   * @return MpackVersion instance which contains parsed version
   * */
  public static MpackVersion parseStackVersion(String stackVersion) {
    Matcher versionMatcher = validateStackVersion(stackVersion);

    if(versionMatcher.pattern().pattern().equals(LEGACY_STACK_VERSION_PATTERN)) {
      return new MpackVersion(Integer.parseInt(versionMatcher.group(1)),
        Integer.parseInt(versionMatcher.group(2)),
        Integer.parseInt(versionMatcher.group(3)), Integer.parseInt(versionMatcher.group(4)),
        Integer.parseInt(versionMatcher.group(5)));
    } else {
      throw new IllegalArgumentException("Wrong format for mpack version");
    }
  }

  /**
   * Method validate stack version not to be
   * empty or null. Also check if passed version
   * has valid format.
   * @param version string
   * @return Matcher for passed version
   * @throws IllegalArgumentException() if version empty/null/not valid
   */
  private static Matcher validateStackVersion(String version) {
    if (StringUtils.isEmpty(version)) {
      throw new IllegalArgumentException("Stack version can't be empty or null");
    }

    String stackVersion = StringUtils.trim(version);

    Matcher versionMatcher = PATTERN_LEGACY_STACK_VERSION.matcher(stackVersion);
    if (!versionMatcher.find()) {
      throw new IllegalArgumentException("Wrong format for stack version, should be N.N.N.N-N or N.N.N-hN-bN");
    }

    return versionMatcher;
  }

  /**
   * Method validate mpack version not to be
   * empty or null. Also check if passed version
   * has valid format.
   * @param version string
   * @return Matcher for passed version
   * @throws IllegalArgumentException() if version empty/null/not valid
   */
  private static Matcher validateMpackVersion(String version) {
    if (StringUtils.isEmpty(version)) {
      throw new IllegalArgumentException("Mpack version can't be empty or null");
    }

    String mpackVersion = StringUtils.trim(version);

    Matcher versionMatcher = PATTERN_WITH_HOTFIX.matcher(mpackVersion);
    if (!versionMatcher.find()) {
      versionMatcher = PATTERN_WITHOUT_HOTFIX.matcher(mpackVersion);
      if (!versionMatcher.find()) {
        throw new IllegalArgumentException("Wrong format for mpack version, should be N.N.N-bN or N.N.N-hN-bN");
      }
    }

    return versionMatcher;
  }

  @Override
  public int compareTo(MpackVersion other) {
    int result = this.major - other.major;
    if(result == 0) {
      result = this.minor - other.minor;
      if(result == 0) {
        result = this.maint - other.maint;
        if(result == 0) {
          result = this.hotfix - other.hotfix;
          if(result == 0) {
            result = this.build - other.build;
          }
        }
      }
    }
    return result > 0 ? 1 : result < 0 ? -1 : 0;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;

    MpackVersion that = (MpackVersion) o;

    if (build != that.build) return false;
    if (hotfix != that.hotfix) return false;
    if (maint != that.maint) return false;
    if (major != that.major) return false;
    if (minor != that.minor) return false;

    return true;
  }

  @Override
  public int hashCode() {
    int result = major;
    result = 31 * result + minor;
    result = 31 * result + maint;
    result = 31 * result + hotfix;
    result = 31 * result + build;
    return result;
  }
}
